src/inlineSuggestions/stateTracker.ts

```diff
@@ -23,9 +30,11 @@ export function initTracker(): Disposable[] {
   return [
     workspace.onDidChangeTextDocument(
       ({ contentChanges }: TextDocumentChangeEvent) => {
-        const contentChange = contentChanges[0];
+        const contentChange = new Change(contentChanges[0]);
         const changeHappened =
-          contentChange?.rangeLength >= 0 && contentChange?.text !== "";
+          contentChange.isValid() &&
+          contentChange.isNotIndentationChange() &&
+          contentChange.isSingleCharChange();
```

Comment:
there's a problem with this condition:
- in `isValid` you're assetring that `this.contentChange?.text.trim().length >= 0`
- in `isSingleCharChange` you're asserting that `this.contentChange?.text.trim().length <= 1`
==> you're asserting that `this.contentChange?.text.trim().length == 1`

this means that `isNotIndentationChange` will always evaluate to true, since it asserts that the text is not <a string that's length is `tabSize`> - but you already know that this string's length is exactly 1.
I'm not sure what you wanted to check here, I presume it should be only `isSingleCharChange`, or only `isNotIndentationChange` 🤷🏻‍♂️

Reply:
not exactly...
the `isValid` assets untrimmed value - meaning it was an actual change (and not deletion for example)
the `isSingleCharChange` assert that the "timed" length is 1  - meaning it will allow any single not whitespace char
the `isNotIndentationChange` asserts a non-trimmed value - allowing everything different from the indent

I will rename the function names to better express the intention

Reply:
oh so you allow either [at least one non-whitespace char] or [only whitespace chars, but exactly `tabSize` length]
yeah that's a sneaky one - maybe worth playing around with the functions and conditions to make it clearer 👍🏻